File: programs/benchmarks-alioth/zig/mandelbrot.zig
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// Zig port for p2w comparison (Zig 0.15+)

const std = @import("std");
const posix = std.posix;

fn print(comptime fmt: []const u8, args: anytype) void {
    var buf: [256]u8 = undefined;
    const msg = std.fmt.bufPrint(&buf, fmt, args) catch unreachable;
    _ = posix.write(posix.STDOUT_FILENO, msg) catch {};
}

pub fn main() !void {
    var args = std.process.args();
    _ = args.skip();

    const size_str = args.next() orelse "200";
    const size: i32 = std.fmt.parseInt(i32, size_str, 10) catch 200;

    const w = size;
    const h = size;
    const iter: i32 = 50;
    const limit: f64 = 2.0;
    const limit_sq = limit * limit;

    var total_in_set: i32 = 0;

    var y: i32 = 0;
    while (y < h) : (y += 1) {
        var x: i32 = 0;
        while (x < w) : (x += 1) {
            const cr = 2.0 * @as(f64, @floatFromInt(x)) / @as(f64, @floatFromInt(w)) - 1.5;
            const ci = 2.0 * @as(f64, @floatFromInt(y)) / @as(f64, @floatFromInt(h)) - 1.0;

            var zr: f64 = 0.0;
            var zi: f64 = 0.0;
            var tr: f64 = 0.0;
            var ti: f64 = 0.0;

            var i: i32 = 0;
            while (i < iter and (tr + ti <= limit_sq)) : (i += 1) {
                zi = 2.0 * zr * zi + ci;
                zr = tr - ti + cr;
                tr = zr * zr;
                ti = zi * zi;
            }

            if (tr + ti <= limit_sq) {
                total_in_set += 1;
            }
        }
    }

    print("size: {d} in_set: {d}\n", .{ w, total_in_set });
}

File: programs/benchmarks-alioth/python/mandelbrot.py
# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org/
#
# contributed by Tupteq
# modified for single-threaded p2w
#
# This version outputs to stdout but doesn't produce actual PBM image
# It focuses on the computation for benchmarking purposes


def main(size: int) -> None:
    fsize: float = float(size)

    # Just count iterations (for benchmarking without I/O)
    total_in_set: int = 0

    y: i32 = 0
    while y < size:
        fy: float = 2.0 * y / fsize - 1.0

        x: i32 = 0
        while x < size:
            # Complex number c = (cr, ci)
            cr: float = 2.0 * x / fsize - 1.5
            ci: float = fy

            # z = 0
            zr: float = 0.0
            zi: float = 0.0

            # Iterate z = z*z + c
            iter: i32 = 0
            in_set: int = 1
            while iter < 50:
                # z*z = (zr + zi*i)^2 = zr^2 - zi^2 + 2*zr*zi*i
                zr2: float = zr * zr
                zi2: float = zi * zi

                # Check if |z|^2 > 4
                if zr2 + zi2 > 4.0:
                    in_set = 0
                    iter = 50  # break

                # z = z*z + c
                zi = 2.0 * zr * zi + ci
                zr = zr2 - zi2 + cr

                iter = iter + 1

            total_in_set = total_in_set + in_set
            x = x + 1

        y = y + 1

    # Output total count for verification
    print("size:", size, "in_set:", total_in_set)


# Default argument for testing
main(200)
